### PR TITLE
feat: implement functionalities for importing new farms in the farm page

### PR DESCRIPTION
**Pull request by gsantini-ube**

### CODE REVIEW

File: src/components/earn/ImportFarmModal.tsx [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+  const [error, setError] = useState<string | undefined>(undefined)
+
+  const farmExists = isAddress(farmAddress)
+    ? farmSummaries.find(
+        (farmSummary) => farmAddress && getAddress(farmSummary.stakingAddress) === getAddress(farmAddress)
+      )
+    : undefined
+
+  function wrappedOndismiss() {
```
Comment (BriungRi) [inline src/components/earn/ImportFarmModal.tsx:46]: Why is this needed?

File: src/components/earn/ImportFarmModal.tsx [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    onDismiss()
+  }
+
+  const handleInput = useCallback((event) => {
+    const input = event.target.value
+    setFarmAddress(input)
+  }, [])
+
+  useEffect(() => {
```
Comment (BriungRi) [inline src/components/earn/ImportFarmModal.tsx:55]: It would be a good idea to debounce this

File: src/components/earn/ImportFarmModal.tsx [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    if (isAddress(farmAddress)) {
+      const importedFarms = localStorage.getItem('imported_farms')
```
Comment (BriungRi) [inline src/components/earn/ImportFarmModal.tsx:57]: store 'imported_farms' as a constant

File: src/components/earn/ImportFarmModal.tsx [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+            <Text fontWeight={500} fontSize={16}>
+              Import Farm
+            </Text>
+            <CloseIcon onClick={wrappedOndismiss} />
+          </RowBetween>
+          <Row>
+            <SearchInput
+              type="text"
+              id="token-search-input"
+              placeholder={'Input or Paste Farm Address'}
```
Comment (BriungRi) [inline src/components/earn/ImportFarmModal.tsx:95]: Enter farm address

File: src/components/earn/ImportFarmModal.tsx [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+                      })
+                    : '-'}
+                </Text>
+              </RowBetween>
+            </AutoColumn>
+            <AutoColumn justify="center">
+              <RowBetween align={'start'}>
+                <TYPE.black>Rewards Token{rewardTokens && rewardTokens.length > 0 ? 's' : ''}</TYPE.black>
+                <AutoColumn justify="end">
+                  {rewardTokens ? (
```
Comment (BriungRi) [inline src/components/earn/ImportFarmModal.tsx:131]: Unnecessary to show `rewardTokens`

File: src/components/earn/PoolCard.tsx [hunk 6]
```diff
@@ -121,33 +149,82 @@ export const PoolCard: React.FC<Props> = ({ farmSummary }: Props) => {
...
     return null
   }
 
   return (
-    <Wrapper showBackground={isStaking} bgColor={theme.primary1}>
+    <Wrapper showBackground={isStaking} bgColor={farmSummary.isImported ? theme.bg5 : theme.primary1}>
+      <RemoveFarmModal
+        isOpen={showRemoveModal}
+        onClose={() => setShowRemoveModal(false)}
+        onConfirm={() => onRemoveFarm()}
```
Comment (BriungRi) [inline src/components/earn/PoolCard.tsx:184]: {onRemoveFarm}

**Review by BriungRi — CHANGES_REQUESTED** [review]
Please stop modifying the .husky/pre-commit file. I think there may be software on your computer that is modifying file permissions.

File: src/components/earn/ImportFarmModal.tsx [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+  const [error, setError] = useState<string | undefined>(undefined)
+
+  const farmExists = isAddress(farmAddress)
+    ? farmSummaries.find(
+        (farmSummary) => farmAddress && getAddress(farmSummary.stakingAddress) === getAddress(farmAddress)
+      )
+    : undefined
+
+  const importedFarms = localStorage.getItem('imported_farms')
```
Comment (BriungRi) [inline src/components/earn/ImportFarmModal.tsx:46]: store 'imported_farms' as a constant

File: src/components/earn/ImportFarmModal.tsx [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    const input = event.target.value
+    setFarmAddress(input)
+  }
+
+  useEffect(() => {
+    if (isAddress(farmAddress)) {
+      const res = importedFarms
+        ? [...JSON.parse(importedFarms)].find((item) => getAddress(item) === getAddress(farmAddress))
+        : undefined
+      setError(res ? 'The farm has already been imported' : undefined)
```
Comment (BriungRi) [inline src/components/earn/ImportFarmModal.tsx:58]: Store this message in en.json

File: src/components/earn/ImportFarmModal.tsx [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+        setError('Enter valid farm address')
+      } else {
+        setError(undefined)
+      }
+    }
+  }, [farmAddress, importedFarms])
+
+  const onConfirm = () => {
+    localStorage.setItem(
+      'imported_farms',
```
Comment (BriungRi) [inline src/components/earn/ImportFarmModal.tsx:70]: use constant

File: src/components/earn/ImportFarmModal.tsx [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+      JSON.stringify(importedFarms ? [...JSON.parse(importedFarms), farmAddress] : [farmAddress])
```
Comment (BriungRi) [inline src/components/earn/ImportFarmModal.tsx:71]: What happens if `importedFarms` is ever a malformed JSON?

E.g. try `JSON.parse('af8= 9a8')`

We should catch instances of this where we could potentially crash the app and gracefully reset the local storage if we detect that this happens.

Comment (BriungRi) [reply src/components/earn/ImportFarmModal.tsx:71]: Also, what happens if any of the farms in `importedFarms` is maliciously updated to not be a farm? Do we handle this case gracefully too?

File: src/components/earn/ImportFarmModal.tsx [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+            !!error ||
+            !!farmExists ||
+            !valueOfTotalStakedAmountInCUSD ||
+            !rewardTokens ||
+            !totalRewardRates ||
+            !stakingToken
+          }
+          onClick={onConfirm}
+        >
+          {error ? error : farmExists ? 'The Farm Already Exists' : 'Import Farm'}
```
Comment (BriungRi) [inline src/components/earn/ImportFarmModal.tsx:157]: Use en.json

File: src/components/earn/PoolCard.tsx [hunk 12]
```diff
@@ -39,7 +42,7 @@ const StatContainer = styled.div`
 const Wrapper = styled(AutoColumn)<{ showBackground: boolean; bgColor: any }>`
   border-radius: 12px;
   width: 100%;
-  overflow: hidden;
+  overflow: unset;
```
Comment (BriungRi) [inline src/components/earn/PoolCard.tsx:45]: why was this added?

File: src/components/earn/PoolCard.tsx [hunk 13]
```diff
@@ -121,33 +149,82 @@ export const PoolCard: React.FC<Props> = ({ farmSummary }: Props) => {
...
+      {farmSummary.isImported && !isStaking && (
+        <RemoveCardSection>
+          <CloseIcon onClick={() => setShowRemoveModal(true)} />
+        </RemoveCardSection>
+      )}
       <CardNoise />
 
-      <TopSection>
-        <DoubleCurrencyLogo currency0={token0} currency1={token1} size={24} />
+      <TopSection singleToken={!!token0 && !!token1 && token0.address === token1.address ? true : false}>
```
Comment (BriungRi) [inline src/components/earn/PoolCard.tsx:193]: nit: dont need `true : false`

File: src/components/earn/PoolCard.tsx [hunk 14]
```diff
@@ -121,33 +149,82 @@ export const PoolCard: React.FC<Props> = ({ farmSummary }: Props) => {
...
+        {token0 && token1 && token0.address === token1.address ? (
```
Comment (BriungRi) [inline src/components/earn/PoolCard.tsx:194]: store 
```
const singleToken = token0 && token1 && token0.address === token1.address
```

File: src/components/earn/PoolCard.tsx [hunk 15]
```diff
@@ -158,11 +235,12 @@ export const PoolCard: React.FC<Props> = ({ farmSummary }: Props) => {
                 </TYPE.small>
               </TYPE.white>
             </span>
-          )}
+          ) : null}
         </PoolInfo>
-
         <StyledInternalLink
-          to={`/farm/${token0?.address}/${token1?.address}/${farmSummary.stakingAddress}`}
+          to={`/farm/${
```
Comment (BriungRi) [inline src/components/earn/PoolCard.tsx:241]: just have two paths for single staking vs. multi staking

File: src/components/earn/RemoveFarmModal.tsx [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  return (
+    <Modal isOpen={isOpen} onDismiss={onDismiss} maxHeight={100}>
+      <ModalContentWrapper>
+        <AutoColumn gap="lg">
+          <RowBetween style={{ padding: '0 2rem' }}>
+            <div />
+            <Text fontWeight={500} fontSize={20}>
+              Are you sure?
```
Comment (BriungRi) [inline src/components/earn/RemoveFarmModal.tsx:53]: use en.json where raw string / text is being displayed

File: src/constants/homoraBank.ts [hunk 17]
```diff
@@ -0,0 +1,8 @@
+import { ChainId } from '@ubeswap/sdk'
```
Comment (BriungRi) [inline src/constants/homoraBank.ts:1]: Why are we adding this to this PR?

File: src/pages/Earn/index.tsx [hunk 18]
```diff
@@ -39,27 +42,88 @@ const Header: React.FC = ({ children }) => {
...
+      setPrevImportedFarms(importedFarms)
+      setVisibleImportedFarms(importedFarms)
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [importedFarms])
+
+  const handleRemoveFarm = (farmAddress: string) => {
+    const _importedFarms = visibleImportedFarms.filter((farm) => farm.stakingAddress !== farmAddress)
+    setVisibleImportedFarms(_importedFarms)
+    const farms_saved = localStorage.getItem('imported_farms')
```
Comment (BriungRi) [inline src/pages/Earn/index.tsx:120]: use constant

File: src/pages/Earn/index.tsx [hunk 19]
```diff
@@ -39,27 +42,88 @@ const Header: React.FC = ({ children }) => {
...
+    if (
+      !(
+        importedFarms.length === prevImportedFarms.length &&
+        importedFarms.every((value, index) => value === prevImportedFarms[index])
+      )
+    ) {
+      setPrevImportedFarms(importedFarms)
+      setVisibleImportedFarms(importedFarms)
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (BriungRi) [inline src/pages/Earn/index.tsx:114]: why is this necessary?

File: src/pages/Earn/index.tsx [hunk 20]
```diff
@@ -90,9 +154,51 @@ export default function Earn() {
         </TopSection>
       )}
       <TopSection gap="md">
-        <AutoColumn>
-          <TokenSelect onTokenSelect={setFilteringToken} token={filteringToken} />
-        </AutoColumn>
+        <MobileContainer>
+          {' '}
```
Comment (BriungRi) [inline src/pages/Earn/index.tsx:158]: nit: kill line

File: src/pages/Earn/useFarmRegistry.ts [hunk 21]
```diff
@@ -105,19 +134,201 @@ export const useFarmRegistry = () => {
   return farmSummaries
 }
 
-export const useUniqueBestFarms = () => {
-  const farmSummaries = useFarmRegistry()
+export const useImportedFarms = () => {
+  const importedFarmsAddress = localStorage.getItem('imported_farms')
```
Comment (BriungRi) [inline src/pages/Earn/useFarmRegistry.ts:138]: use constant

File: src/pages/Earn/useFarmRegistry.ts [hunk 22]
```diff
@@ -105,19 +134,201 @@ export const useFarmRegistry = () => {
...
+  const tokens = useAllTokens()
+  const library = useProvider()
+  const provider = getProviderOrSigner(library, account ? account : undefined)
+  const cusd = cUSD[chainId as unknown as UbeswapChainId]
+  const bank = useMemo(
+    () => new ethers.Contract(Bank[chainId], BANK_ABI.abi as ContractInterface, provider) as unknown as HomoraBank,
+    [chainId, provider]
+  )
+
+  // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (BriungRi) [inline src/pages/Earn/useFarmRegistry.ts:156]: is this ignore statement necessary?

File: src/pages/Earn/useFarmRegistry.ts [hunk 23]
```diff
@@ -105,19 +134,201 @@ export const useFarmRegistry = () => {
...
+        const [externalRewardsToken, rewardRate] = await Promise.all([
+          moolaStaking.rewardsToken(),
+          moolaStaking.rewardRate(),
+        ])
+        tokens.push(externalRewardsToken)
+        rates.push(rewardRate)
+        if (i < externalEarned.length - 1) stakingRewardsAddress = await moolaStaking.externalStakingRewards()
+      }
+    } catch (err) {
+      // console.log(err)
```
Comment (BriungRi) [inline src/pages/Earn/useFarmRegistry.ts:183]: uncomment

File: src/pages/Earn/useFarmRegistry.ts [hunk 24]
```diff
@@ -105,19 +134,201 @@ export const useFarmRegistry = () => {
...
+
+  const getPairToken = async (pair: ethers.Contract): Promise<PairToken | undefined> => {
+    let token0Address: string | undefined = undefined
+    let token1Address: string | undefined = undefined
+    try {
+      const tokens = await Promise.all([pair.token0(), pair.token1()])
+      token0Address = tokens[0]
+      token1Address = tokens[1]
+    } catch (err) {
+      // console.log(err)
```
Comment (BriungRi) [inline src/pages/Earn/useFarmRegistry.ts:196]: uncomment. Use console.error

File: src/components/earn/PoolCard.tsx (hunk 12 above)
Comment (gsantini-ube) [reply src/components/earn/PoolCard.tsx:45]: image
To show the whole close button on the pool card, I've changed it and it doesn't affect other features or styles for the pool card.

**Comment by gsantini-ube** [comment]
> Why are ABIs like CoreOracle, HomoraBank being added to this PR?
image
image

To calculated amount of total deposited, I needed to get celo price of staking token.

File: src/components/earn/ImportFarmModal.tsx [hunk 25]
```diff
@@ -0,0 +1,169 @@
...
+        : undefined
+      setError(res ? t('TheFarmHasAlreadyBeenImported') : undefined)
+    } else {
+      if (farmAddress.length > 0) {
+        setError(t('EnterValidFarmAddress'))
+      } else {
+        setError(undefined)
+      }
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (BriungRi) [inline src/components/earn/ImportFarmModal.tsx:67]: what are we suppressing here?

File: src/components/earn/ImportFarmModal.tsx [hunk 26]
```diff
@@ -0,0 +1,169 @@
...
+
+  const handleInput = (event: any) => {
+    const input = event.target.value
+    setFarmAddress(input)
+  }
+
+  useEffect(() => {
+    if (isAddress(farmAddress)) {
+      const res = importedFarms
+        ? [...JSON.parse(importedFarms)].find((item) => getAddress(item) === getAddress(farmAddress))
```
Comment (BriungRi) [inline src/components/earn/ImportFarmModal.tsx:57]: why are you doing `[...]`?

Why can't you just do `JSON.parse(importedFarms).find..`

File: src/components/earn/ImportFarmModal.tsx [hunk 27]
```diff
@@ -0,0 +1,169 @@
...
+      }
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [farmAddress, importedFarms])
+
+  const onConfirm = () => {
+    try {
+      localStorage.setItem(
+        IMPORTED_FARMS,
+        JSON.stringify(importedFarms ? [...JSON.parse(importedFarms), farmAddress] : [farmAddress])
```
Comment (BriungRi) [inline src/components/earn/ImportFarmModal.tsx:74]: ditto. Why `[...]`?

File: src/components/earn/ImportedPoolCard.tsx [hunk 28]
```diff
@@ -0,0 +1,21 @@
...
+
+export const ImportedPoolCard: React.FC<Props> = ({ farmAddress, onUpdateFarm }: Props) => {
+  const farmSummary = useImportedFarmRegistry(farmAddress)
+  const [prevFarmSummary, setPrevFarmSummary] = useState<FarmSummary | undefined>(undefined)
+  useEffect(() => {
+    if (farmSummary && !_.isEqual(farmSummary, prevFarmSummary)) {
+      setPrevFarmSummary(farmSummary)
+      onUpdateFarm(farmSummary)
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (BriungRi) [inline src/components/earn/ImportedPoolCard.tsx:18]: what are we suppressing?

File: src/components/earn/ImportedPoolCard.tsx [hunk 29]
```diff
@@ -0,0 +1,21 @@
+import _ from 'lodash'
```
Comment (BriungRi) [inline src/components/earn/ImportedPoolCard.tsx:1]: don't import `_`. Import `isEqual` directly from `lodash`

File: src/components/earn/PoolCard.tsx [hunk 30]
```diff
@@ -121,33 +149,84 @@ export const PoolCard: React.FC<Props> = ({ farmSummary }: Props) => {
...
-  try {
-    compoundedAPY = annualizedPercentageYield(apr, COMPOUNDS_PER_YEAR)
-  } catch (e) {
-    // eslint-disable-next-line @typescript-eslint/no-unused-vars
-    console.error('apy calc overflow', farmSummary.farmName, e)
+  if (!farmSummary.isImported) {
+    try {
+      compoundedAPY = annualizedPercentageYield(apr, COMPOUNDS_PER_YEAR)
+    } catch (e) {
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (BriungRi) [inline src/components/earn/PoolCard.tsx:156]: why are we disabling eslint here?

File: src/components/earn/PoolCard.tsx (hunk 14 above)
Comment (BriungRi) [reply src/components/earn/PoolCard.tsx:196]: unresolved still

File: src/constants/abis/IUniswapV2Pair.json [hunk 31]
```diff
@@ -0,0 +1,653 @@
+[
```
Comment (BriungRi) [inline src/constants/abis/IUniswapV2Pair.json:1]: Unnecessary

You can just use
```
import IUniswapV2Pair from '@ubeswap/core/build/abi/IUniswapV2Pair.json'

```

See src/constants/abis/uniswap-v2-pair.ts

File: src/components/earn/ImportFarmModal.tsx (hunk 27 above)
Comment (gsantini-ube) [reply src/components/earn/ImportFarmModal.tsx:73]: If we do things like [JSON.parse(importedFarms), farmAddress], we will get [[importedFarmAddress1, importedFarmAddress2, ...], farmAddress] while we need to get [addr1, addr2, ...]

Comment (BriungRi) [reply src/components/earn/ImportFarmModal.tsx:73]: It is probably faster to do the following:

importedFarms ? JSON.parse(importedFarms).concat(farmAddress) : [farmAddress]

File: public/locales/en.json [hunk 32]
```diff
@@ -230,6 +230,13 @@
   "youHaveUnclaimedRewards": "You have {{count}} farms with unclaimed rewards",
   "claimAllRewards": "Claim all rewards",
   "featuredPools": "Featured Pools",
+  "importedPools": "Imported Pools",
+  "ImportFarm": "Import Farm",
+  "TheFarmAlreadyExists": "The Farm Already Exists",
+  "EnterFarmAddress": "Enter farm address",
+  "EnterValidFarmAddress": "Enter valid farm address",
+  "TheFarmHasAlreadyBeenImported": "The farm has already been imported",
+  "AreYouSure": "Are you sure?"
```
Comment (BriungRi) [inline public/locales/en.json:239]: Missing comma

**Review by BriungRi — APPROVED** [review]
